Server should send possible_teleporters

This is like possible_placements, but will be keyed on heister/color: possible_teleporters { yellow: [mp_a, mp_b, mp_c], green: [mp_a]}

If a color/heister color is not present, then there are no valid teleports for that heister.

I didn't have to do this before, but I wasn't defining new list entry messages before - 
```
error[E0277]: the trait bound `types::proto_types::PossibleTeleportEntry: endpoints::_IMPL_DESERIALIZE_FOR_CreateGameFormData::_serde::Serialize` is not satisfied
   --> /Users/kemcbride/github/team_heist_tactics/target/debug/build/team_heist_tactics-295488de4f141a52/out/types.rs:134:5
    |
134 | /     /// Map of possible locations a heister can teleport to
135 | |     #[prost(message, repeated, tag="12")]
136 | |     pub possible_teleports: ::prost::alloc::vec::Vec<PossibleTeleportEntry>,
    | |___________________________________________________________________________^ the trait `endpoints::_IMPL_DESERIALIZE_FOR_CreateGameFormData::_serde::Serialize` is not implemented for `types::proto_types::PossibleTeleportEntry`
    |
    = note: required because of the requirements on the impl of `endpoints::_IMPL_DESERIALIZE_FOR_CreateGameFormData::_serde::Serialize` for `std::vec::Vec<types::proto_types::PossibleTeleportEntry>`
    = note: required by `endpoints::_IMPL_DESERIALIZE_FOR_CreateGameFormData::_serde::ser::SerializeStruct::serialize_field`
```

--- a/build.rs
+++ b/build.rs
@@ -21,6 +21,10 @@ fn main() {
         .type_attribute(".types.TilePosition", "#[derive(Serialize, Deserialize)]")
         .type_attribute(".types.HeisterName", "#[derive(Serialize, Deserialize)]")
         .type_attribute(".types.HeisterColor", "#[derive(Serialize, Deserialize)]")
+        .type_attribute(
+            ".types.PossibleTeleportEntry",
+            "#[derive(Serialize, Deserialize)]",
+        )
         .type_attribute(".types.StartGame", "#[derive(Serialize, Deserialize)]")
         .compile_protos(&["src/types.proto"], &["src/"])
         .unwrap();

--- a/src/game.rs
+++ b/src/game.rs
@@ -25,6 +25,7 @@ pub struct Game {
     pub game_state: GameState,
     pub tile_deck: Vec<Tile>,
     pub game_created: u64,
+    revealed_teleporters: HashMap<HeisterColor, Vec<MapPosition>>,
 }
 #[derive(Clone, Default, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
 pub struct GameHandle(pub String);
@@ -56,11 +57,18 @@ impl Game {
             tile_deck.shuffle(&mut rng);
         }
         let game_created = get_current_time_secs();
+        // NOTE: Assumption: All games start with only one tile revealed
+        let mut revealed_teleporters: HashMap<HeisterColor, Vec<MapPosition>> = HashMap::new();
+        Self::update_revealed_teleporters(
+            &mut revealed_teleporters,
+            game_state.tiles.first().unwrap(),
+        );
         Game {
             game_handle,
             game_state,
             tile_deck,
             game_created,
+            revealed_teleporters,
         }
     }
 
@@ -478,13 +486,14 @@ impl Game {
     }
 
     fn update_auxiliary_state(&mut self) -> () {
-        self.update_possible_placements();
-        self.update_possible_escalators();
+        let grid = self.get_absolute_grid();
+        self.update_possible_placements(&grid);
+        self.update_possible_escalators(&grid);
+        self.update_possible_teleports(&grid);
     }
 
     /// Possible placements for new tiles that Heisters can discover
-    fn update_possible_placements(&mut self) -> () {
-        let grid = self.get_absolute_grid();
+    fn update_possible_placements(&mut self, grid: &HashMap<MapPosition, Square>) -> () {
         let heister_to_tile_entrance_locs = self.heister_to_tile_entrance_positions(&grid);
 
         let mut v = Vec::new();
@@ -495,8 +504,7 @@ impl Game {
     }
 
     /// Possible escalator destinations that a Heister can reach with an Escalator move
-    fn update_possible_escalators(&mut self) -> () {
-        let grid = self.get_absolute_grid();
+    fn update_possible_escalators(&mut self, grid: &HashMap<MapPosition, Square>) -> () {
         let mut m: HashMap<HeisterColor, MapPosition> = HashMap::new();
         for heister in &self.game_state.heisters {
             let color = heister.heister_color;
@@ -510,6 +518,45 @@ impl Game {
             }
         }
         self.game_state.possible_escalators = m;
+    }
+
+    /// Possible teleport destinations that a Heister can reach with a Teleport move
+    fn update_possible_teleports(&mut self, grid: &HashMap<MapPosition, Square>) -> () {
+        let mut m: HashMap<HeisterColor, Vec<MapPosition>> = HashMap::new();
+        for heister in &self.game_state.heisters {
+            let color = heister.heister_color;
+            let pos = &heister.map_position;
+            let square = grid.get(&pos).unwrap();
+            if square.is_teleport() {
+                match self.revealed_teleporters.get(&color) {
+                    Some(list) => {
+                        m.insert(color, list.to_vec());
+                    }
+                    None => {}
+                }
+            }
+        }
+        self.game_state.possible_teleports = m;
+    }
+
+    /// Return all revealed teleporters, called after adding new tiles
+    fn update_revealed_teleporters(
+        already_revealed: &mut HashMap<HeisterColor, Vec<MapPosition>>,
+        new_tile: &Tile,
+    ) -> () {
+        for (color, teleporter_pos) in new_tile.get_teleporters() {
+            match already_revealed.get_mut(&color) {
+                Some(already_revealed_list) => {
+                    // append the new list onto the end
+                    already_revealed_list.push(teleporter_pos);
+                }
+                None => {
+                    // set the new list from teleporter list
+                    let teleporter_list: Vec<MapPosition> = vec![teleporter_pos];
+                    already_revealed.insert(color, teleporter_list);
+                }
+            }
+        }
     }
 
     /// From a tile entrance and move direction of the tile's orientation,
@@ -686,8 +733,9 @@ impl Game {
                 // Add the tile before opening doors on it, that way helpers that
                 // rely on the tile's presence in game.tiles work correctly
                 let new_tile_idx = self.game_state.tiles.len();
+                Self::update_revealed_teleporters(&mut self.revealed_teleporters, &rotated_tile);
                 self.game_state.tiles.push(rotated_tile);
-                self.update_tile_doors(new_tile_idx);
+                self.update_tile_doors(new_tile_idx); // Must be called _after_ push
                 MoveValidity::Valid
             }
             None => MoveValidity::Invalid("No tiles left in deck to draw".to_string()),

--- a/src/types.proto
+++ b/src/types.proto
@@ -132,6 +132,13 @@ enum GameStatus {
   DEFEAT = 3;
 }
 
+// We can't have maps of lists in protobuf, so instead we'll have a list of these,
+// and construct a map from that.
+message PossibleTeleportEntry {
+    HeisterColor color = 1;
+    MapPosition position = 2;
+}
+
 // Captures all information related to the game.
 message GameState {
   // Word(s) that points to this game instance.
@@ -166,6 +173,9 @@ message GameState {
 
   // Possible escalator positions a heister can reach via escalator
   map<int32, MapPosition> possible_escalators = 11;
+
+  // Map of possible locations a heister can teleport to
+  repeated PossibleTeleportEntry possible_teleports = 12;
 
 }
 

--- a/src/types.rs
+++ b/src/types.rs
@@ -16,6 +16,7 @@ pub use proto_types::Ability;
 pub use proto_types::GameStatus;
 pub use proto_types::HeisterColor;
 pub use proto_types::HeisterSymbol;
+pub use proto_types::PossibleTeleportEntry;
 pub use proto_types::SquareType;
 pub use proto_types::WallType;
 
@@ -38,6 +39,13 @@ pub const DOOR_TYPES: [&'static WallType; 4] = [
     &WallType::OrangeDoor,
     &WallType::GreenDoor,
     &WallType::YellowDoor,
+];
+
+pub const TELEPORTER_TYPES: [&'static SquareType; 4] = [
+    &SquareType::PurpleTeleportPad,
+    &SquareType::OrangeTeleportPad,
+    &SquareType::GreenTeleportPad,
+    &SquareType::YellowTeleportPad,
 ];
 
 pub const HEISTER_COLORS: [&'static HeisterColor; 4] = [
@@ -411,6 +419,26 @@ impl Tile {
         }
         None
     }
+
+    /// Returns a Map from color to MapPositions for all teleporters within the tile
+    /// NOTE: Assumption: Each tile has no more than 1 teleporter per color
+    pub fn get_teleporters(&self) -> HashMap<HeisterColor, MapPosition> {
+        let mut map: HashMap<HeisterColor, MapPosition> = HashMap::new();
+        for (idx, square) in self.squares.iter().enumerate() {
+            if !TELEPORTER_TYPES.contains(&&square.square_type) {
+                continue;
+            }
+            // Match teleporter type to color,... helper???
+            match square.color() {
+                Some(color) => {
+                    map.insert(color, self.square_idx_to_map_pos(idx));
+                }
+                None => {}
+            }
+        }
+        // fn square_idx_to_map_pos(&self, i: usize) -> MapPosition {
+        map
+    }
 }
 
 pub enum StartingTile {
@@ -464,6 +492,16 @@ impl From<SerializableSquare> for Square {
 }
 
 impl Square {
+    pub fn color(&self) -> Option<HeisterColor> {
+        match self.square_type {
+            SquareType::PurpleTeleportPad => Some(HeisterColor::Purple),
+            SquareType::OrangeTeleportPad => Some(HeisterColor::Orange),
+            SquareType::GreenTeleportPad => Some(HeisterColor::Green),
+            SquareType::YellowTeleportPad => Some(HeisterColor::Yellow),
+            _wildcard => None,
+        }
+    }
+
     pub fn is_teleport(&self) -> bool {
         match self.square_type {
             SquareType::PurpleTeleportPad
@@ -758,6 +796,7 @@ pub struct GameState {
     pub players: Vec<Player>,
     pub possible_placements: Vec<MapPosition>,
     pub possible_escalators: HashMap<HeisterColor, MapPosition>,
+    pub possible_teleports: HashMap<HeisterColor, Vec<MapPosition>>,
 }
 
 impl Internal for GameState {
@@ -796,6 +835,22 @@ impl Internal for GameState {
                 )
             })
             .collect();
+
+        // Have to actually process this list into a proper map
+        let mut possible_teleports: HashMap<HeisterColor, Vec<MapPosition>> = HashMap::new();
+        for entry in proto.possible_teleports {
+            let color = HeisterColor::from_i32(entry.color).unwrap();
+            let pos = MapPosition::from_proto(entry.position.unwrap());
+            match possible_teleports.get_mut(&color) {
+                Some(list) => {
+                    list.push(pos);
+                }
+                None => {
+                    let list: Vec<MapPosition> = vec![pos];
+                    possible_teleports.insert(color, list);
+                }
+            }
+        }
         GameState {
             game_name,
             game_started: proto.game_started,
@@ -808,6 +863,7 @@ impl Internal for GameState {
             players,
             possible_placements,
             possible_escalators,
+            possible_teleports,
         }
     }
 
@@ -826,6 +882,17 @@ impl Internal for GameState {
             .map(|(c, pe)| (i32::from(*c), pe.to_proto()))
             .collect();
         let game_status = i32::from(self.game_status);
+
+        let mut possible_teleports: Vec<PossibleTeleportEntry> = Vec::new();
+        for (color, list) in self.possible_teleports.clone() {
+            for pos in list {
+                let entry = PossibleTeleportEntry {
+                    color: i32::from(color),
+                    position: Some(pos.to_proto()),
+                };
+                possible_teleports.push(entry);
+            }
+        }
         proto_types::GameState {
             game_name: self.game_name.0.to_string(),
             game_started: self.game_started,
@@ -838,6 +905,7 @@ impl Internal for GameState {
             players,
             possible_placements,
             possible_escalators,
+            possible_teleports,
         }
     }
 }
@@ -850,6 +918,7 @@ impl GameState {
         let starting_tile_enum = StartingTile::A(starting_tile.clone());
         let tiles = vec![starting_tile.clone()];
         let possible_escalators = HashMap::new();
+        let possible_teleports = HashMap::new();
         let mut heisters = Vec::new();
         heisters.push(Heister::get_initial(
             HeisterColor::Yellow,
@@ -879,6 +948,7 @@ impl GameState {
             players: vec![],
             possible_placements: vec![],
             possible_escalators,
+            possible_teleports,
         }
     }
 }

